crm_stage_type/models/crm_lead.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2018 Eficent Business and IT Consulting Services, S.L.
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import SUPERUSER_ID, api, fields, models
+
+
+class Lead(models.Model):
+    _inherit = "crm.lead"
+
+    stage_id = fields.Many2one(
+        domain="[('team_id', 'in', [team_id, False]), "
+        "('lead_type', 'in', [type, 'both'])]"
+    )
+
+    @api.model
+    def _read_group_stage_ids(self, stages, domain, order):
+        ctx_type = self.env.context.get("default_type")
+        stages = super(Lead, self)._read_group_stage_ids(stages, domain, order)
+        search_domain = [("id", "in", stages.ids)]
+        if ctx_type:
+            search_domain += [("lead_type", "in", [ctx_type, "both"])]
+        stage_ids = stages._search(
+            search_domain, order=order, access_rights_uid=SUPERUSER_ID
+        )
+        return stages.browse(stage_ids)
+
+    def _stage_find(self, team_id=False, domain=None, order="sequence"):
+        # check whether we should try to add a condition on type
+        domain = domain or []
+        if not any(
+            [term for term in domain if len(term) == 3 if term[0] == "lead_type"]
```

Comment:
why use 2 if's? you could use 'and'.
`[term for term in domain if len(term) == 3 and term[0] == "lead_type"]`